### PR TITLE
add pester for retriable client

### PR DESCRIPTION
**Pull request by kfangw**
### Github Issue
#878 

### Background
see #878 

### Solution
adopt https://github.com/sethgrid/pester
Retry strategy is configurable

### CODE REVIEW

File: lib/common/http2_client.go [hunk 1]
```diff
@@ -38,14 +43,24 @@ func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client
...
+		Timeout:   timeout,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse // NOTE prevent redirect
 		},
+	}
+
+	ec := pester.NewExtendedClient(hc)
+	{
+		ec.Concurrency = 1
+		ec.MaxRetries = 5
```
[the remark this answers is not in the document]
Comment (Charleslee522) [reply lib/common/http2_client.go]: I think greater than 10 is better if there is not an infinite option.

[the remark this answers is not in the document]
Comment (spikeekips) [reply lib/common/http2_client.go]: We also need to make a network client without retry, so the `*pester.Client` can be set in argument of `NewHTTP2Client`.

**Comment by kfangw** [comment]
@spikeekips add an argument for retry
@Charleslee522 add unit test. I think it is enough.
Please review again ~

**Comment by kfangw** [comment]
> @kfangw Looks good. One thing I suggest is, it will be better to be able to set the pester.Client in NewHTTP2Client directly. There are various requirements to change the retry count and concurrency, etc .

I thought about that, but at this time, it is enough I think.
I will make another PR later to enhance this to be configurable.

File: lib/common/http2_client.go [hunk 2]
```diff
@@ -11,12 +12,16 @@ import (
 	"golang.org/x/net/http2"
 )
 
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
 type HTTP2Client struct {
-	client    http.Client
+	client    HTTPClient
```
Comment (anarcher) [inline lib/common/http2_client.go]: I feel that `HTTPClient` in `HTTP2Client` is awful. :->  How about `Doer` or `HTTPDoer`?

File: lib/common/http2_client.go [hunk 3]
```diff
@@ -38,14 +43,29 @@ func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client
...
+		Timeout:   timeout,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse // NOTE prevent redirect
 		},
+	}
+	httpClient = hc
+
+	if retry {
+		ec := pester.NewExtendedClient(hc)
+		ec.Concurrency = 1
```
Comment (anarcher) [inline lib/common/http2_client.go]: Can I configure this options for another purpose? For example, with syncer, retry options can be different.

Comment (kfangw) [reply lib/common/http2_client.go]: OK then, I will make this configurable.

**Comment by kfangw** [comment]
@anarcher Could you review the latest commit again? dedf502

File: lib/common/http2_client.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package common
 import (
 	"bytes"
 	"crypto/tls"
+	"github.com/sethgrid/pester"
```
Comment (Geod24) [inline lib/common/http2_client.go]: Please order import

File: lib/network/http2.go [hunk 5]
```diff
@@ -130,7 +130,18 @@ func NewHTTP2Network(config *HTTP2NetworkConfig) (h2n *HTTP2Network) {
 
 // GetClient creates new keep-alive HTTP2 client
 func (t *HTTP2Network) GetClient(endpoint *common.Endpoint) NetworkClient {
-	rawClient, _ := common.NewHTTP2Client(defaultTimeout, 0, true)
+	rawClient, _ := common.NewHTTP2Client(
```
Comment (Geod24) [inline lib/network/http2.go]: We shouldn't be ignoring errors here

File: lib/common/http2_client.go [hunk 6]
```diff
@@ -38,14 +50,29 @@ func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client
...
+	var doer Doer
+
+	hc := &http.Client{
+		Transport: transport,
+		Timeout:   timeout,
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse // NOTE prevent redirect
 		},
+	}
+	doer = hc
```
Comment (Geod24) [inline lib/common/http2_client.go]: Can't you assign to `doer` directly ?

File: lib/common/http2_client.go [hunk 7]
```diff
@@ -38,14 +50,29 @@ func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client
...
+
+	if retrySetting != nil {
+		ec := pester.NewExtendedClient(hc)
+		{
+			ec.MaxRetries = retrySetting.MaxRetries
+			ec.Concurrency = retrySetting.Concurrency
+			ec.Backoff = retrySetting.Backoff
+		}
+		doer = ec
```
Comment (Geod24) [inline lib/common/http2_client.go]: Same here

File: lib/network/http2_test.go [hunk 8]
```diff
@@ -4,9 +4,11 @@ import (
 	"crypto/tls"
 	"errors"
 	"fmt"
+	"github.com/gorilla/mux"
```
Comment (Geod24) [inline lib/network/http2_test.go]: Order imports

File: lib/common/http2_client.go [hunk 9]
```diff
@@ -11,12 +12,23 @@ import (
...
+}
+
 type HTTP2Client struct {
-	client    http.Client
+	client    Doer
 	transport *http.Transport
 }
 
-func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool) (client *HTTP2Client, err error) {
+func NewHTTP2Client(timeout, idleTimeout time.Duration, keepAlive bool, retrySetting *RetrySetting) (client *HTTP2Client, err error) {
```
Comment (Geod24) [inline lib/common/http2_client.go]: Given the amount of modification necessary, I'd suggest to provide another overload. E.g. leave `NewHTTP2Client` with the current interface, and make it forward to `NewPersistentHTTP2Client` with `nil`

**Comment by kfangw** [comment]
@anarcher @Geod24 Please review again.

**Pull request merged**